OpenRA.Mods.Common/Orders/BeaconOrderGenerator.cs

```diff
@@ -30,7 +31,8 @@ protected override IEnumerable<Order> OrderInner(World world, CPos cell, int2 wo
 		protected override IEnumerable<IRenderable> RenderAnnotations(WorldRenderer wr, World world) { yield break; }
 		protected override string GetCursor(World world, CPos cell, int2 worldPixel, MouseInput mi)
 		{
-			return "ability"; // TODO: [CursorReference]
+			// Always return the beacon cursor as the command cannot be blocked.
+			return world.LocalPlayer.PlayerActor.Info.TraitInfo<PlaceBeaconInfo>().BeaconCursor;
```

Comment:
How do I get "world" at that point though? I could also look up the trait on the first call and store the result.

Reply:
You don't need it. Follow the example of `SelectConditionTarget` and all the other existing order generators.

Reply:
Oh, I see this is coming from order buttons, so there is no trait invoking this. Still, you have `World` in `BeaconOrderButtonLogic` so you can cache there. You need to find a way to make `BindOrderButton` accept parameters.

Reply:
Maybe caching the trait inside the `GetCursor()` using `CachedTransform` (as it will be moved to OpenRA.Game.Primitives soon) would be good enough solution and wouldn't require changing `BeaconOrderButtonLogic`.The key would be either `Player` or `World`.

I personally would expect beacon cursor to be defined on the trait and not on a Logic object buried in chrome YAML. Many other traits define cursors for their feature in their info classes.

Reply:
I don't think `CachedTransform` is the right thing for a cursor reference that never changes once the rules are loaded.

Reply:
You're right, `World.ToggleInputMode<T>()` creates new `IOrderGenerator` each time it is changed, so there's no point in caching it in the `BeaconOrderGenerator`.

Reply:
Would caching it inside the BeaconOrderGenerator Class be okay?

`string beaconCursor;
protected override string GetCursor(World world, CPos cell, int2 worldPixel, MouseInput mi)
{
	// Always return the beacon cursor as the command cannot be blocked.
	if (beaconCursor == "")
		beaconCursor = world.LocalPlayer.PlayerActor.Info.TraitInfo<PlaceBeaconInfo>().BeaconCursor;
	return beaconCursor;
}`

Reply:
You can cache it in the constructor of `BeaconOrderButtonLogic` instead.

Reply:
I'm trying to figure out a satisfying way to get the cursor into the order generator this way. All I could come up with so far is passing a method to `BindOrderButton` which is invoked on the world's OrderGenerator after `OnClick` sets it. This seems a rather specific change to `BindOrderButton` though, I'd prefer a more flexible approach. Anyone else got ideas?